src/apiutil.h

```diff
@@ -308,3 +316,208 @@ bool hasInsufficientMaterial(Color c, const Position& pos) {
 
     return true;
 }
+
+enum FEN_VALIDATION : int {
+  FEN_MISSING_SPACE_DELIM = -12,
+  FEN_INVALID_NB_PARTS = -11,
+  FEN_INVALID_CHAR = -10,
+  FEN_FILE_SIZE_MISMATCH = -9,
+  FEN_INVALID_NUMBER_OF_RANKS = -8,
+  FEN_INVALID_POCKET_INFO = -7,
+  FEN_INVALID_SIDE_TO_MOVE = -6,
+  FEN_INVALID_CASTLING_INFO = -5,
+  FEN_INVALID_EN_PASSANT_SQ = -4,
+  FEN_INVALID_CHECK_COUNTER = -3,
+  FEN_INVALID_HALF_MOVE_COUNTER = -2,
+  FEN_INVALID_MOVE_COUTNER = -1,
+  FEN_EMPTY = 0,
+  FEN_OK = 1
+};
+
+
+FEN_VALIDATION validate_fen(const std::string& fen, const Variant* v) {
+
+  // 0) Layout
+  // check for empty fen
+  if (fen.size() == 0) {
+    std::cerr << "Fen is empty." << std::endl;
+    return FEN_EMPTY;
+  }
+
+  // check for space
+  if (fen.find(' ') == std::string::npos) {
+    std::cerr << "Fen misses space as delimiter." << std::endl;
+    return FEN_MISSING_SPACE_DELIM;
+  }
+
+  std::vector<std::string> fenParts;
+  std::vector<std::string> starFenParts;
+  std::stringstream ss(fen);
+  std::stringstream ss2(v->startFen);
+
+  std::string curPart;
+  while (std::getline(ss, curPart, ' '))
+    fenParts.emplace_back(curPart);
+  while (std::getline(ss2, curPart, ' '))
+    starFenParts.emplace_back(curPart);
+  const unsigned int nbFenParts = starFenParts.size();
+
+  // check for number of parts
+  if (fenParts.size() != nbFenParts) {
+    std::cerr << "Invalid number of fen parts. Expected: " << nbFenParts << " Actual: " << fenParts.size() << std::endl;
+    return FEN_INVALID_NB_PARTS;
+  }
+
+  // 1) Part
+  // check for valid characters
+  for (char c : fenParts[0])
+    if (!isdigit(c) && v->startFen.find(c) == std::string::npos) {
+      std::cerr << "Invalid piece character: '" << c << "'." << std::endl;
+      return FEN_INVALID_CHAR;
+    }
+
+  // check for number of ranks
+  int nbRanks = RANK_1;
+  for (; nbRanks <= v->maxRank; ++nbRanks);
+  // check for number of files
+  int nbFiles = FILE_A;
+  for (; nbFiles <= v->maxFile; ++nbFiles);
+
+  int nbProcessedRanks = 1;  // start with 1 and increment after delimeter
+  int curRankWidth = 0;
+  char prevChar = '?';
+  for (char c : fenParts[0]) {
+    if (isdigit(c)) {
+      curRankWidth += c - '0';
+      // if previous digit exists we have a dual digit
+      if (isdigit(prevChar))
+        curRankWidth += 9;
```

Comment:
This is a bit hacky to me, I needed to read it several times to understand that the additional digit is first counted as 1 and therefore only 9 is added in the next iteration. If it can not easily be solved otherwise, then I think at least writing it as `10 * (prevChar - '0') - 1` makes it clearer, and as an aside also fixes more artifical cases like 01 or 20.

Reply:
I agree that this formulation is a bit hard to understand.
Using `9 * (prevChar - '0')` allows board sizes up to **99x99**.
```C++
curRankWidth += 9 * (prevChar - '0');
```
Alternative, we could not iterate over single characters and convert the number using **stoi()**.